runtime/runtime/src/actions.rs

```diff
@@ -192,10 +192,22 @@ pub(crate) fn action_function_call(
         false,
     );
     let execution_succeeded = match err {
-        Some(VMError::FunctionCallError(err)) => {
-            result.result = Err(ActionErrorKind::FunctionCallError(err).into());
-            false
-        }
+        Some(VMError::FunctionCallError(err)) => match err {
```

Comment:
That's intentional, and for that particular distinction we can introduce CompilationError and ExecutionError, if this difference is indeed important. Currently, the only use of the error is showing it to developers in logs.

Reply:
> Currently, the only use of the error is showing it to developers in logs.

This is exactly what we want to improve by introducing sane granularity.

> That's intentional, and for that particular distinction we can introduce CompilationError and ExecutionError, if this difference is indeed important.

Yes, that would be ideal. I think end-users care about three types of errors: UnknownMethod, CompilationError, ExecutionError.

Reply:
Sounds good, please introduce these 3 levels of errors.

Reply:
Cool, let's do that!

Reply:
+1 for structured errors, it's one of the pain points in `near-api-js` and all projects that are using it.
What do you think about adding the `BadArgument` error? It's a common mistake for developers.

Reply:
@volovyk-s  `BadArgument` is a contract-specific error (because each contract decides to interpret it differently). We cannot differentiate it on a blockchain-level.

Reply:
@olonho It seems like we are still combining all three kinds of errors into `ActionErrorKind::FunctionCallError` with string. Did you forget to push the commit?

Reply:
> and we prefer to have error in RPC as structured as possible

I want to push back a little bit against "as structured as possible errors". This is not strictly related to the PR, just the  pattern I observer in Rust's error handling. 

The issue with _maximally_ stuctured errors is that they increase coupling, and sometimes it's better to stop at the goldilocks errors, which are structured just enough. 

In the context of this PR, if we make error maximally structured, than we need to expose wasmer 0.17 and wasmer 1.0 as is. And when we remove support for wasmer 0.17 or upgrade to (hypothetical) wasmer 2.0, we'd have to change the definition of the error type as well, and update the call-sites. This PR in some sense is the consequence of the fact that we did change error definition when upgrading to wasmer 1.0. This is a code smell, pointing out that the original  error definition probably wasn't encapsulated enough. 

But, clearly, exposing *just* the error message goes to far in the opposite direction, as it forces the clients to parse out the messages. 

There's a goldilocks error pattern -- `ErrorKind`. The best example of it is std::io::Error:

![image](https://user-images.githubusercontent.com/1711539/113852042-c75b1d00-97a4-11eb-8a78-7001028de5f7.png)

The `Error` type itself is an opaque struct, which can be converted to string via `Display`. There's a `.kind()` method, which exposes **non-exhaustive**, **field-less** enum. This structure allows the client to behave differently for different sorts of errors, but allows the implementor to modify the errors  without changes to the client. 

To re-iterate, this is definitely not something we should do in the scope of this PR, and it's not necessary something we want to do even eventually (serializability of errors add an extra design dimension here which might change the tradeoffs). But it's useful to keep this bit of info in mind -- just exposing errors as is is the path of least resistance, but often not something you want to have at the API _boundary_. 

References: 

* https://github.com/rust-lang-nursery/failure/blob/master/book/src/error-errorkind.md
* https://matklad.github.io/2020/10/15/study-of-std-io-error.html
* https://blog.burntsushi.net/rust-error-handling/

Reply:
@matklad thanks for the detailed overview. I believe after some discussions we already arrived at the middle ground of introducing exactly the necessary variants that the consumer cares about :+1: